PolyTypeGenerator throws unhandled exception

I'm trying to migrate one of my MessagPack-consuming projects (closed source) to Nerdbank.MessagePack. I got a ways through, but now I'm hitting this that's blocking further progress:

> 4>CSC : warning CS8785: Generator 'PolyTypeGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'GenerateShapeProvider.IncomingViewingKey.g.cs' of the added source file must be unique within a generator.

--- a/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
+++ b/src/PolyType.SourceGenerator/Helpers/RoslynHelpers.cs
@@ -152,20 +152,25 @@ internal static partial class RoslynHelpers
     }
 
     /// <summary>
-    /// Returns a string representation of the type suitable for use as an identifier in source.
+    /// Returns a string representation of the type suitable for use as an identifier in source code or file names.
     /// </summary>
-    public static string CreateTypeIdentifier(this ITypeSymbol type)
+    public static string CreateTypeIdentifier(this ITypeSymbol type, ReadOnlySpan<string> reservedIdentifiers, bool includeNamespaces = false)
     {
         StringBuilder sb = new();
         GenerateCore(type, sb);
-        return sb.ToString();
+        string identifier = sb.ToString();
 
-        static void GenerateCore(ITypeSymbol type, StringBuilder sb)
+        // Do not return identifiers that are C# keywords or reserved identifiers.
+        return IsCSharpKeyword(identifier) || reservedIdentifiers.IndexOf(identifier) >= 0 
+            ? "__Type_" + identifier 
+            : identifier;
+
+        void GenerateCore(ITypeSymbol type, StringBuilder sb)
         {
             switch (type)
             {
                 case ITypeParameterSymbol typeParameter:
-                    AppendAsPascalCase(typeParameter.Name);
+                    sb.Append(typeParameter.Name);
                     break;
 
                 case IArrayTypeSymbol arrayType:
@@ -180,8 +185,13 @@ internal static partial class RoslynHelpers
                     break;
 
                 case INamedTypeSymbol namedType:
-                    PrependContainingTypes(namedType);
-                    AppendAsPascalCase(namedType.Name);
+                    if (includeNamespaces)
+                    {
+                        PrependNamespaces(namedType.ContainingNamespace);
+                        PrependContainingTypes(namedType);
+                    }
+
+                    sb.Append(namedType.Name);
 
                     IEnumerable<ITypeSymbol> typeArguments = namedType.IsTupleType
                         ? namedType.TupleElements.Select(e => e.Type)
@@ -200,6 +210,16 @@ internal static partial class RoslynHelpers
                     throw new InvalidOperationException();
             }
 
+            void PrependNamespaces(INamespaceSymbol ns)
+            {
+                if (ns.ContainingNamespace is { } containingNs)
+                {
+                    PrependNamespaces(containingNs);
+                    sb.Append(ns.Name);
+                    sb.Append('_');
+                }
+            }
+
             void PrependContainingTypes(INamedTypeSymbol namedType)
             {
                 if (namedType.ContainingType is { } parent)
@@ -208,14 +228,6 @@ internal static partial class RoslynHelpers
                     GenerateCore(parent, sb);
                     sb.Append('_');
                 }
-            }
-
-            void AppendAsPascalCase(string name)
-            {
-                // Avoid creating identifiers that are C# keywords
-                Debug.Assert(name.Length > 0);
-                sb.Append(char.ToUpperInvariant(name[0]));
-                sb.Append(name, 1, name.Length - 1);
             }
         }
     }

--- a/src/PolyType.SourceGenerator/Model/TypeId.cs
+++ b/src/PolyType.SourceGenerator/Model/TypeId.cs
@@ -8,7 +8,6 @@ namespace PolyType.SourceGenerator.Model;
 public readonly struct TypeId : IEquatable<TypeId>
 {
     public required string FullyQualifiedName { get; init; }
-    public required string TypeIdentifier { get; init; }
     public required bool IsValueType { get; init; }
     public required SpecialType SpecialType { get; init; }
 

--- a/src/PolyType.SourceGenerator/Model/TypeShapeModel.cs
+++ b/src/PolyType.SourceGenerator/Model/TypeShapeModel.cs
@@ -5,4 +5,9 @@ namespace PolyType.SourceGenerator.Model;
 public abstract record TypeShapeModel
 {
     public required TypeId Type { get; init; }
+
+    /// <summary>
+    /// A unique identifier corresponding to the type that is a valid C# identifier.
+    /// </summary>
+    public required string SourceIdentifier { get; init; }
 }

--- a/src/PolyType.SourceGenerator/Model/TypeShapeModel.cs
+++ b/src/PolyType.SourceGenerator/Model/TypeShapeModel.cs
@@ -7,7 +7,7 @@ public abstract record TypeShapeModel
     public required TypeId Type { get; init; }
 
     /// <summary>
-    /// A unique identifier corresponding to the type that is a valid C# identifier.
+    /// A unique identifier deriving from the type name that can be used as a valid member identifier.
     /// </summary>
     public required string SourceIdentifier { get; init; }
 }

--- a/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
@@ -7,28 +7,28 @@ namespace PolyType.SourceGenerator;
 
 public sealed partial class Parser
 {
-    private TypeShapeModel MapModel(TypeDataModel model, TypeId typeId, string propertyIdentifier)
+    private TypeShapeModel MapModel(TypeDataModel model, TypeId typeId, string sourceIdentifier)
     {
         return model switch
         {
             EnumDataModel enumModel => new EnumShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 UnderlyingType = CreateTypeId(enumModel.UnderlyingType),
             },
 
             NullableDataModel nullableModel => new NullableShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 ElementType = CreateTypeId(nullableModel.ElementType),
             },
 
             EnumerableDataModel enumerableModel => new EnumerableShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 ElementType = CreateTypeId(enumerableModel.ElementType),
                 ConstructionStrategy = enumerableModel.ConstructionStrategy switch
                 {
@@ -67,7 +67,7 @@ public sealed partial class Parser
             DictionaryDataModel dictionaryModel => new DictionaryShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 KeyType = CreateTypeId(dictionaryModel.KeyType),
                 ValueType = CreateTypeId(dictionaryModel.ValueType),
                 ConstructionStrategy = dictionaryModel.ConstructionStrategy switch
@@ -105,7 +105,7 @@ public sealed partial class Parser
             ObjectDataModel objectModel => new ObjectShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 Constructor = objectModel.Constructors
                     .Select(c => MapConstructor(objectModel, typeId, c))
                     .FirstOrDefault(),
@@ -123,7 +123,7 @@ public sealed partial class Parser
             TupleDataModel tupleModel => new ObjectShapeModel
             {
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 Constructor = MapTupleConstructor(typeId, tupleModel),
                 Properties = tupleModel.Elements
                     .Select((e, i) => MapProperty(model.Type, typeId, e, tupleElementIndex: i, isClassTupleType: !tupleModel.IsValueTuple))
@@ -137,7 +137,7 @@ public sealed partial class Parser
             _ => new ObjectShapeModel
             { 
                 Type = typeId,
-                SourceIdentifier = propertyIdentifier,
+                SourceIdentifier = sourceIdentifier,
                 Constructor = null,
                 Properties = [],
                 IsValueTupleType = false,

--- a/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
@@ -7,25 +7,28 @@ namespace PolyType.SourceGenerator;
 
 public sealed partial class Parser
 {
-    private TypeShapeModel MapModel(TypeId typeId, TypeDataModel model)
+    private TypeShapeModel MapModel(TypeDataModel model, TypeId typeId, string propertyIdentifier)
     {
         return model switch
         {
             EnumDataModel enumModel => new EnumShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 UnderlyingType = CreateTypeId(enumModel.UnderlyingType),
             },
 
             NullableDataModel nullableModel => new NullableShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 ElementType = CreateTypeId(nullableModel.ElementType),
             },
 
             EnumerableDataModel enumerableModel => new EnumerableShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 ElementType = CreateTypeId(enumerableModel.ElementType),
                 ConstructionStrategy = enumerableModel.ConstructionStrategy switch
                 {
@@ -64,6 +67,7 @@ public sealed partial class Parser
             DictionaryDataModel dictionaryModel => new DictionaryShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 KeyType = CreateTypeId(dictionaryModel.KeyType),
                 ValueType = CreateTypeId(dictionaryModel.ValueType),
                 ConstructionStrategy = dictionaryModel.ConstructionStrategy switch
@@ -101,6 +105,7 @@ public sealed partial class Parser
             ObjectDataModel objectModel => new ObjectShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 Constructor = objectModel.Constructors
                     .Select(c => MapConstructor(objectModel, typeId, c))
                     .FirstOrDefault(),
@@ -118,6 +123,7 @@ public sealed partial class Parser
             TupleDataModel tupleModel => new ObjectShapeModel
             {
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 Constructor = MapTupleConstructor(typeId, tupleModel),
                 Properties = tupleModel.Elements
                     .Select((e, i) => MapProperty(model.Type, typeId, e, tupleElementIndex: i, isClassTupleType: !tupleModel.IsValueTuple))
@@ -131,6 +137,7 @@ public sealed partial class Parser
             _ => new ObjectShapeModel
             { 
                 Type = typeId,
+                SourceIdentifier = propertyIdentifier,
                 Constructor = null,
                 Properties = [],
                 IsValueTupleType = false,

--- a/src/PolyType.SourceGenerator/Parser/Parser.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.cs
@@ -128,27 +128,19 @@ public sealed partial class Parser : TypeDataModelGenerator
 
         Parser parser = new(knownSymbols.Compilation.Assembly, knownSymbols, cancellationToken);
         ImmutableEquatableArray<TypeDeclarationModel> generateShapeTypes = parser.IncludeTypesUsingGenerateShapeAttributes(generateShapeDeclarations);
-        return parser.ExportTypeShapeProviderModel(s_globalImplicitProviderDeclaration, generateShapeTypes);
+        return parser.ExportTypeShapeProviderModel(s_globalProviderDeclaration, generateShapeTypes);
     }
 
     private TypeShapeProviderModel ExportTypeShapeProviderModel(TypeDeclarationModel providerDeclaration, ImmutableEquatableArray<TypeDeclarationModel> generateShapeTypes)
     {
-        Dictionary<TypeId, TypeShapeModel> generatedModels = new(GeneratedModels.Count);
-        foreach (KeyValuePair<ITypeSymbol, TypeDataModel> entry in GeneratedModels)
-        {
-            TypeId typeId = CreateTypeId(entry.Value.Type);
-            if (generatedModels.ContainsKey(typeId))
-            {
-                ReportDiagnostic(TypeNameConflict, location: null, typeId.FullyQualifiedName);
-            }
-
-            generatedModels[typeId] = MapModel(typeId, entry.Value);
-        }
-
         return new TypeShapeProviderModel
         {
             ProviderDeclaration = providerDeclaration,
-            ProvidedTypes = generatedModels.ToImmutableEquatableDictionary(),
+            ProvidedTypes = GetGeneratedTypesAndIdentifiers()
+                .ToImmutableEquatableDictionary(
+                    keySelector: kvp => kvp.Key,
+                    valueSelector: kvp => MapModel(kvp.Value.Model, kvp.Value.TypeId, kvp.Value.SourceIdentifier)),
+
             AnnotatedTypes = generateShapeTypes,
             Diagnostics = Diagnostics.ToImmutableEquatableSet(),
         };
@@ -269,12 +261,63 @@ public sealed partial class Parser : TypeDataModelGenerator
         }
     }
 
+    private Dictionary<TypeId, (TypeDataModel Model, TypeId TypeId, string SourceIdentifier)> GetGeneratedTypesAndIdentifiers()
+    {
+        Dictionary<TypeId, (TypeDataModel Model, TypeId TypeId, string SourceIdentifier)> results = new(GeneratedModels.Count);
+        Dictionary<string, TypeId?> shortIdentifiers = new(GeneratedModels.Count);
+        ReadOnlySpan<string> reservedIdentifiers = SourceFormatter.ReservedIdentifiers;
+
+        foreach (KeyValuePair<ITypeSymbol, TypeDataModel> entry in GeneratedModels)
+        {
+            TypeId typeId = CreateTypeId(entry.Value.Type);
+            if (results.ContainsKey(typeId))
+            {
+                // We can't have duplicate types with the same fully qualified name.
+                ReportDiagnostic(TypeNameConflict, location: null, typeId.FullyQualifiedName);
+                continue;
+            }
+
+            // Generate a property name for the type. Start with a short-form name that
+            // doesn't include namespaces or containing types. If there is a conflict,
+            // we will update the identifiers to incorporate fully qualified names.
+            // Fully qualified names should not have conflicts since we've already checked
+
+            string sourceIdentifier = entry.Value.Type.CreateTypeIdentifier(reservedIdentifiers, includeNamespaces: false);
+            if (!shortIdentifiers.TryGetValue(sourceIdentifier, out TypeId? conflictingIdentifier))
+            {
+                // This is the first occurrence of the short-form identifier.
+                // Add to the index including the typeId in case of a later conflict.
+                shortIdentifiers.Add(sourceIdentifier, typeId);
+            }
+            else
+            {
+                // We have a conflict, update the identifiers of both types to long-form.
+                if (conflictingIdentifier is { } cId)
+                {
+                    // Update the identifier of the conflicting type since it hasn't been already.
+                    var conflictingResults = results[cId];
+                    conflictingResults.SourceIdentifier = conflictingResults.Model.Type.CreateTypeIdentifier(reservedIdentifiers, includeNamespaces: true);
+                    results[cId] = conflictingResults;
+
+                    // Mark the short-form identifier as updated.
+                    shortIdentifiers[sourceIdentifier] = null;
+                }
+
+                // Update the identifier of the current type and store the new key.
+                sourceIdentifier = entry.Value.Type.CreateTypeIdentifier(reservedIdentifiers, includeNamespaces: true);
+            }
+
+            results.Add(typeId, (entry.Value, typeId, sourceIdentifier));
+        }
+
+        return results;
+    }
+
     private static TypeId CreateTypeId(ITypeSymbol type)
     {
         return new TypeId
         {
             FullyQualifiedName = type.GetFullyQualifiedName(),
-            TypeIdentifier = type.CreateTypeIdentifier(),
             IsValueType = type.IsValueType,
             SpecialType = type.OriginalDefinition.SpecialType,
         };
@@ -290,12 +333,11 @@ public sealed partial class Parser : TypeDataModelGenerator
         return null;
     }
 
-    private static readonly TypeDeclarationModel s_globalImplicitProviderDeclaration = new()
+    private static readonly TypeDeclarationModel s_globalProviderDeclaration = new()
     {
         Id = new()
         {
             FullyQualifiedName = "global::PolyType.SourceGenerator.ShapeProvider",
-            TypeIdentifier = "GenerateShapeProvider",
             IsValueType = false,
             SpecialType = SpecialType.None,
         },

--- a/src/PolyType.SourceGenerator/PolyTypeGenerator.cs
+++ b/src/PolyType.SourceGenerator/PolyTypeGenerator.cs
@@ -42,7 +42,7 @@ public sealed class PolyTypeGenerator : IIncrementalGenerator
             context.ReportDiagnostic(diagnostic.CreateDiagnostic());
         }
 
-        SourceFormatter.FormatProvider(context, provider);
+        SourceFormatter.GenerateSourceFiles(context, provider);
     }
 
     public Action<TypeShapeProviderModel>? OnGeneratingSource { get; init; }

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Constructors.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Constructors.cs
@@ -8,15 +8,14 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private const string BitArrayFQN = "global::System.Collections.BitArray";
     private const string FlagsArgumentStateLabel = "Flags";
 
-    private static void FormatConstructorFactory(SourceWriter writer, string methodName, ObjectShapeModel type, ConstructorShapeModel constructor)
+    private void FormatConstructorFactory(SourceWriter writer, string methodName, ObjectShapeModel type, ConstructorShapeModel constructor)
     {
         string constructorArgumentStateFQN = FormatConstructorArgumentStateFQN(type, constructor);
-        string? constructorParameterFactoryName = constructor.TotalArity > 0 ? FormatConstructorParameterFactoryName(type) : null;
+        string? constructorParameterFactoryName = constructor.TotalArity > 0 ? $"__CreateConstructorParameters_{type.SourceIdentifier}" : null;
         
         writer.WriteLine($"private global::PolyType.Abstractions.IConstructorShape {methodName}()");
         writer.WriteLine('{');
@@ -25,7 +24,7 @@ internal static partial class SourceFormatter
         writer.WriteLine($$"""
             return new global::PolyType.SourceGenModel.SourceGenConstructorShape<{{type.Type.FullyQualifiedName}}, {{constructorArgumentStateFQN}}>
             {
-                DeclaringType = (global::PolyType.Abstractions.IObjectTypeShape<{{type.Type.FullyQualifiedName}}>){{type.Type.TypeIdentifier}},
+                DeclaringType = (global::PolyType.Abstractions.IObjectTypeShape<{{type.Type.FullyQualifiedName}}>){{type.SourceIdentifier}},
                 ParameterCount = {{constructor.TotalArity}},
                 GetParametersFunc = {{FormatNull(constructorParameterFactoryName)}},
                 DefaultConstructorFunc = {{FormatDefaultCtor(type, constructor)}},
@@ -85,7 +84,7 @@ internal static partial class SourceFormatter
                         .Select(FormatDefaultValueExpr)
                         .Append(
                             flagType is OptionalMemberFlagsType.BitArray 
-                            ? $"new {BitArrayFQN}({constructor.OptionalMembers.Length})"
+                            ? $"new({constructor.OptionalMembers.Length})"
                             : "default")),
             };
 
@@ -150,12 +149,13 @@ internal static partial class SourceFormatter
                     _ => $$""" { {{FormatInitializerBody()}} }""",
                 };
 
+                string constructorName = FormatConstructorName(type, constructor);
                 string constructorExpr = constructor.Parameters.Length switch
                 {
-                    0 when memberInitializerBlock is null => $"{FormatConstructorName(constructor)}()",
-                    0 => $"{FormatConstructorName(constructor)}{memberInitializerBlock}",
-                    1 when constructor.TotalArity == 1 => $"{FormatConstructorName(constructor)}({FormatCtorParameterExpr(constructor.Parameters[0], isSingleParameter: true)})",
-                    _ => $"{FormatConstructorName(constructor)}({FormatCtorArgumentsBody()}){memberInitializerBlock}",
+                    0 when memberInitializerBlock is null => $"{constructorName}()",
+                    0 => $"{constructorName}{memberInitializerBlock}",
+                    1 when constructor.TotalArity == 1 => $"{constructorName}({FormatCtorParameterExpr(constructor.Parameters[0], isSingleParameter: true)})",
+                    _ => $"{constructorName}({FormatCtorArgumentsBody()}){memberInitializerBlock}",
                 };
 
                 // Initialize required members using regular assignments if the constructor is not accessible.
@@ -197,12 +197,12 @@ internal static partial class SourceFormatter
                         string refPrefix = parameter.DeclaringType.IsValueType ? "ref " : "";
                         if (parameter.IsField)
                         {
-                            string accessorName = GetFieldAccessorName(parameter.DeclaringType, parameter.UnderlyingMemberName);
+                            string accessorName = GetFieldAccessorName(type, parameter.UnderlyingMemberName);
                             return $"{accessorName}({refPrefix}obj) = {FormatCtorParameterExpr(parameter)};";
                         }
                         else
                         {
-                            string accessorName = GetPropertySetterAccessorName(parameter.DeclaringType, parameter.UnderlyingMemberName);
+                            string accessorName = GetPropertySetterAccessorName(type, parameter.UnderlyingMemberName);
                             return $"{accessorName}({refPrefix}obj, {FormatCtorParameterExpr(parameter)});";
                         }
                     }
@@ -235,29 +235,26 @@ internal static partial class SourceFormatter
             }
         }
 
-        static string FormatDefaultCtor(ObjectShapeModel type, ConstructorShapeModel constructor)
+        static string FormatDefaultCtor(ObjectShapeModel declaringType, ConstructorShapeModel constructor)
             => constructor.TotalArity switch
             {
-                0 when type.IsValueTupleType => $"static () => default({type.Type.FullyQualifiedName})",
-                0 => $"static () => {FormatConstructorName(constructor)}()",
+                0 when declaringType.IsValueTupleType => $"static () => default({declaringType.Type.FullyQualifiedName})",
+                0 => $"static () => {FormatConstructorName(declaringType, constructor)}()",
                 _ => "null",
             };
 
-        static string FormatConstructorName(ConstructorShapeModel constructor)
+        static string FormatConstructorName(ObjectShapeModel declaringType, ConstructorShapeModel constructor)
         {
             return constructor switch
             {
                 { StaticFactoryName: string factoryName } => factoryName,
-                { IsAccessible: false } => GetConstructorAccessorName(constructor.DeclaringType),
+                { IsAccessible: false } => GetConstructorAccessorName(declaringType),
                 _ => $"new {constructor.DeclaringType.FullyQualifiedName}",
             };
         }
-
-        static string FormatConstructorParameterFactoryName(TypeShapeModel type) =>
-            $"CreateConstructorParameters_{type.Type.TypeIdentifier}";
     }
 
-    private static void FormatConstructorParameterFactory(SourceWriter writer, ObjectShapeModel type, string methodName, ConstructorShapeModel constructor, string constructorArgumentStateFQN)
+    private void FormatConstructorParameterFactory(SourceWriter writer, ObjectShapeModel type, string methodName, ConstructorShapeModel constructor, string constructorArgumentStateFQN)
     {
         writer.WriteLine($"private global::PolyType.Abstractions.IConstructorParameterShape[] {methodName}() => new global::PolyType.Abstractions.IConstructorParameterShape[]");
         writer.WriteLine('{');
@@ -278,7 +275,7 @@ internal static partial class SourceFormatter
                 {
                     Position = {{parameter.Position}},
                     Name = {{FormatStringLiteral(parameter.Name)}},
-                    ParameterType = {{parameter.ParameterType.TypeIdentifier}},
+                    ParameterType = {{GetShapeModel(parameter.ParameterType).SourceIdentifier}},
                     Kind = {{FormatParameterKind(parameter)}},
                     IsRequired = {{FormatBool(parameter.IsRequired)}},
                     IsNonNullable = {{FormatBool(parameter.IsNonNullable)}},
@@ -384,7 +381,7 @@ internal static partial class SourceFormatter
             OptionalMemberFlagsType.UShort => "ushort",
             OptionalMemberFlagsType.UInt32 => "uint",
             OptionalMemberFlagsType.ULong => "ulong",
-            OptionalMemberFlagsType.BitArray => BitArrayFQN,
+            OptionalMemberFlagsType.BitArray => "global::System.Collections.BitArray",
             _ => null,
         };
 
@@ -411,12 +408,12 @@ internal static partial class SourceFormatter
             => $"({string.Join(", ", parameterTypes)})";
     }
 
-    private static string GetConstructorAccessorName(TypeId declaringType)
+    private static string GetConstructorAccessorName(ObjectShapeModel declaringType)
     {
-        return $"{declaringType.TypeIdentifier}_CtorAccessor";
+        return $"__CtorAccessor_{declaringType.SourceIdentifier}";
     }
 
-    private static void FormatConstructorAccessor(SourceWriter writer, ConstructorShapeModel constructorModel)
+    private static void FormatConstructorAccessor(SourceWriter writer, ObjectShapeModel declaringType, ConstructorShapeModel constructorModel)
     {
         Debug.Assert(!constructorModel.IsAccessible);
 
@@ -440,7 +437,7 @@ internal static partial class SourceFormatter
             parameterSignature.Length -= 2;
         }
 
-        string accessorName = GetConstructorAccessorName(constructorModel.DeclaringType);
+        string accessorName = GetConstructorAccessorName(declaringType);
         writer.WriteLine($"""
             [global::System.Runtime.CompilerServices.UnsafeAccessor(global::System.Runtime.CompilerServices.UnsafeAccessorKind.Constructor)]
             private static extern {constructorModel.DeclaringType.FullyQualifiedName} {accessorName}({parameterSignature});

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Dictionary.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Dictionary.cs
@@ -3,17 +3,17 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatDictionaryTypeShapeFactory(SourceWriter writer, string methodName, DictionaryShapeModel dictionaryShapeModel)
+    private void FormatDictionaryTypeShapeFactory(SourceWriter writer, string methodName, DictionaryShapeModel dictionaryShapeModel)
     {
         writer.WriteLine($$"""
             private global::PolyType.Abstractions.ITypeShape<{{dictionaryShapeModel.Type.FullyQualifiedName}}> {{methodName}}()
             {
                 return new global::PolyType.SourceGenModel.SourceGenDictionaryTypeShape<{{dictionaryShapeModel.Type.FullyQualifiedName}}, {{dictionaryShapeModel.KeyType.FullyQualifiedName}}, {{dictionaryShapeModel.ValueType.FullyQualifiedName}}>
                 {
-                    KeyType = {{dictionaryShapeModel.KeyType.TypeIdentifier}},
-                    ValueType = {{dictionaryShapeModel.ValueType.TypeIdentifier}},
+                    KeyType = {{GetShapeModel(dictionaryShapeModel.KeyType).SourceIdentifier}},
+                    ValueType = {{GetShapeModel(dictionaryShapeModel.ValueType).SourceIdentifier}},
                     GetDictionaryFunc = {{FormatGetDictionaryFunc(dictionaryShapeModel)}},
                     ConstructionStrategy = {{FormatCollectionConstructionStrategy(dictionaryShapeModel.ConstructionStrategy)}},
                     DefaultConstructorFunc = {{FormatDefaultConstructorFunc(dictionaryShapeModel)}},

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Enum.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Enum.cs
@@ -3,16 +3,16 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatEnumTypeShapeFactory(SourceWriter writer, string methodName, EnumShapeModel enumShapeType)
+    private void FormatEnumTypeShapeFactory(SourceWriter writer, string methodName, EnumShapeModel enumShapeType)
     {
         writer.WriteLine($$"""
             private global::PolyType.Abstractions.ITypeShape<{{enumShapeType.Type.FullyQualifiedName}}> {{methodName}}()
             {
                 return new global::PolyType.SourceGenModel.SourceGenEnumTypeShape<{{enumShapeType.Type.FullyQualifiedName}}, {{enumShapeType.UnderlyingType.FullyQualifiedName}}>
                 {
-                    UnderlyingType = {{enumShapeType.UnderlyingType.TypeIdentifier}},
+                    UnderlyingType = {{GetShapeModel(enumShapeType.UnderlyingType).SourceIdentifier}},
                     Provider = this,
                 };
             }

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Enumerable.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Enumerable.cs
@@ -3,16 +3,16 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatEnumerableTypeShapeFactory(SourceWriter writer, string methodName, EnumerableShapeModel enumerableShapeModel)
+    private void FormatEnumerableTypeShapeFactory(SourceWriter writer, string methodName, EnumerableShapeModel enumerableShapeModel)
     {
         writer.WriteLine($$"""
             private global::PolyType.Abstractions.ITypeShape<{{enumerableShapeModel.Type.FullyQualifiedName}}> {{methodName}}()
             {
                 return new global::PolyType.SourceGenModel.SourceGenEnumerableTypeShape<{{enumerableShapeModel.Type.FullyQualifiedName}}, {{enumerableShapeModel.ElementType.FullyQualifiedName}}>
                 {
-                    ElementType = {{enumerableShapeModel.ElementType.TypeIdentifier}},
+                    ElementType = {{GetShapeModel(enumerableShapeModel.ElementType).SourceIdentifier}},
                     ConstructionStrategy = {{FormatCollectionConstructionStrategy(enumerableShapeModel.ConstructionStrategy)}},
                     DefaultConstructorFunc = {{FormatDefaultConstructorFunc(enumerableShapeModel)}},
                     EnumerableConstructorFunc = {{FormatEnumerableConstructorFunc(enumerableShapeModel)}},

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.ITypeShapeProvider.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.ITypeShapeProvider.cs
@@ -4,7 +4,7 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
     private static SourceText FormatProviderInterfaceImplementation(TypeShapeProviderModel provider)
     {
@@ -33,7 +33,7 @@ internal static partial class SourceFormatter
         {
             writer.WriteLine($"""
                 if (type == typeof({generatedType.Type.FullyQualifiedName}))
-                    return {generatedType.Type.TypeIdentifier};
+                    return {generatedType.SourceIdentifier};
 
                 """);
         }
@@ -48,7 +48,7 @@ internal static partial class SourceFormatter
         return writer.ToSourceText();
     }
 
-    private static SourceText FormatIShapeableOfTStub(TypeDeclarationModel typeDeclaration, TypeId typeToImplement, TypeShapeProviderModel provider)
+    private SourceText FormatIShapeableOfTStub(TypeDeclarationModel typeDeclaration, TypeId typeToImplement, TypeShapeProviderModel provider)
     {
         var writer = new SourceWriter();
         StartFormatSourceFile(writer, typeDeclaration);
@@ -61,7 +61,7 @@ internal static partial class SourceFormatter
 
         writer.WriteLine($"""
             static global::PolyType.Abstractions.ITypeShape<{typeToImplement.FullyQualifiedName}> global::PolyType.IShapeable<{typeToImplement.FullyQualifiedName}>.GetShape() 
-                => {provider.ProviderDeclaration.Id.FullyQualifiedName}.Default.{typeToImplement.TypeIdentifier};
+                => {provider.ProviderDeclaration.Id.FullyQualifiedName}.{ProviderSingletonProperty}.{GetShapeModel(typeToImplement).SourceIdentifier};
             """);
 
         writer.Indentation--;
@@ -80,7 +80,7 @@ internal static partial class SourceFormatter
             {{typeDeclaration.TypeDeclarationHeader}} : global::PolyType.ITypeShapeProvider
             {
                 global::PolyType.Abstractions.ITypeShape? global::PolyType.ITypeShapeProvider.GetShape(global::System.Type type) 
-                    => {{provider.ProviderDeclaration.Id.FullyQualifiedName}}.Default.GetShape(type);
+                    => {{provider.ProviderDeclaration.Id.FullyQualifiedName}}.{{ProviderSingletonProperty}}.GetShape(type);
             }
             """);
 

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Nullable.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Nullable.cs
@@ -3,9 +3,9 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatNullableTypeShapeFactory(SourceWriter writer, string methodName, NullableShapeModel nullableShapeModel)
+    private void FormatNullableTypeShapeFactory(SourceWriter writer, string methodName, NullableShapeModel nullableShapeModel)
     {
         writer.WriteLine($$"""
             private global::PolyType.Abstractions.ITypeShape<{{nullableShapeModel.Type.FullyQualifiedName}}> {{methodName}}()
@@ -13,7 +13,7 @@ internal static partial class SourceFormatter
                 return new global::PolyType.SourceGenModel.SourceGenNullableTypeShape<{{nullableShapeModel.ElementType.FullyQualifiedName}}>
                 {
                     Provider = this,
-                    ElementType = {{nullableShapeModel.ElementType.TypeIdentifier}},
+                    ElementType = {{GetShapeModel(nullableShapeModel.ElementType).SourceIdentifier}},
                 };
             }
             """);

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Object.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Object.cs
@@ -3,12 +3,12 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatObjectTypeShapeFactory(SourceWriter writer, string methodName, ObjectShapeModel objectShapeModel)
+    private void FormatObjectTypeShapeFactory(SourceWriter writer, string methodName, ObjectShapeModel objectShapeModel)
     {
-        string? propertiesFactoryMethodName = objectShapeModel.Properties.Length > 0 ? $"CreateProperties_{objectShapeModel.Type.TypeIdentifier}" : null;
-        string? constructorFactoryMethodName = objectShapeModel.Constructor != null ? $"CreateConstructor_{objectShapeModel.Type.TypeIdentifier}" : null;
+        string? propertiesFactoryMethodName = objectShapeModel.Properties.Length > 0 ? $"__CreateProperties_{objectShapeModel.SourceIdentifier}" : null;
+        string? constructorFactoryMethodName = objectShapeModel.Constructor != null ? $"__CreateConstructor_{objectShapeModel.SourceIdentifier}" : null;
 
         writer.WriteLine($$"""
             private global::PolyType.Abstractions.ITypeShape<{{objectShapeModel.Type.FullyQualifiedName}}> {{methodName}}()
@@ -46,21 +46,21 @@ internal static partial class SourceFormatter
             if (property.IsField)
             {
                 writer.WriteLine();
-                FormatFieldAccessor(writer, property);
+                FormatFieldAccessor(writer, objectShapeModel, property);
             }
             else
             {
                 if (property is { EmitGetter: true, IsGetterAccessible: false })
                 {
                     writer.WriteLine();
-                    FormatPropertyGetterAccessor(writer, property);
+                    FormatPropertyGetterAccessor(writer, objectShapeModel, property);
                 }
 
                 if (property is { EmitSetter: true, IsSetterAccessible: false } || 
                     (objectShapeModel.Constructor is not null && property.IsInitOnly))
                 {
                     writer.WriteLine();
-                    FormatPropertySetterAccessor(writer, property);
+                    FormatPropertySetterAccessor(writer, objectShapeModel, property);
                 }
             }
         }
@@ -68,7 +68,7 @@ internal static partial class SourceFormatter
         if (objectShapeModel.Constructor is { IsAccessible: false } ctor)
         {
             writer.WriteLine();
-            FormatConstructorAccessor(writer, ctor);
+            FormatConstructorAccessor(writer, objectShapeModel, ctor);
         }
     }
 }

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Properties.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Properties.cs
@@ -6,9 +6,9 @@ using System.Diagnostics;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static void FormatPropertyFactory(SourceWriter writer, string methodName, ObjectShapeModel type)
+    private void FormatPropertyFactory(SourceWriter writer, string methodName, ObjectShapeModel type)
     {
         writer.WriteLine($"private global::PolyType.Abstractions.IPropertyShape[] {methodName}() => new global::PolyType.Abstractions.IPropertyShape[]");
         writer.WriteLine('{');
@@ -40,10 +40,10 @@ internal static partial class SourceFormatter
                 new global::PolyType.SourceGenModel.SourceGenPropertyShape<{{type.Type.FullyQualifiedName}}, {{property.PropertyType.FullyQualifiedName}}>
                 {
                     Name = {{FormatStringLiteral(property.Name)}},
-                    DeclaringType = (global::PolyType.Abstractions.IObjectTypeShape<{{type.Type.FullyQualifiedName}}>){{type.Type.TypeIdentifier}},
-                    PropertyType = {{property.PropertyType.TypeIdentifier}},
-                    Getter = {{(property.EmitGetter ? $"static (ref {type.Type.FullyQualifiedName} obj) => {FormatGetterBody("obj", property)}{(suppressGetter ? "!" : "")}" : "null")}},
-                    Setter = {{(property.EmitSetter ? $"static (ref {type.Type.FullyQualifiedName} obj, {property.PropertyType.FullyQualifiedName} value) => {FormatSetterBody("obj", "value" + (suppressSetter ? "!" : ""), property)}" : "null")}},
+                    DeclaringType = (global::PolyType.Abstractions.IObjectTypeShape<{{type.Type.FullyQualifiedName}}>){{type.SourceIdentifier}},
+                    PropertyType = {{GetShapeModel(property.PropertyType).SourceIdentifier}},
+                    Getter = {{(property.EmitGetter ? $"static (ref {type.Type.FullyQualifiedName} obj) => {FormatGetterBody("obj", type, property)}{(suppressGetter ? "!" : "")}" : "null")}},
+                    Setter = {{(property.EmitSetter ? $"static (ref {type.Type.FullyQualifiedName} obj, {property.PropertyType.FullyQualifiedName} value) => {FormatSetterBody("obj", "value" + (suppressSetter ? "!" : ""), type, property)}" : "null")}},
                     AttributeProviderFunc = {{FormatAttributeProviderFunc(type, property)}},
                     IsField = {{FormatBool(property.IsField)}},
                     IsGetterPublic = {{FormatBool(property.IsGetterPublic)}},
@@ -65,19 +65,19 @@ internal static partial class SourceFormatter
                     : $$"""static () => typeof({{property.DeclaringType.FullyQualifiedName}}).GetProperty({{FormatStringLiteral(property.UnderlyingMemberName)}}, {{InstanceBindingFlagsConstMember}}, null, typeof({{property.PropertyType.FullyQualifiedName}}), [], null)""";
             }
 
-            static string FormatGetterBody(string objParam, PropertyShapeModel property)
+            static string FormatGetterBody(string objParam, ObjectShapeModel declaringType, PropertyShapeModel property)
             {
                 if (!property.IsGetterAccessible)
                 {
                     string refPrefix = property.DeclaringType.IsValueType ? "ref " : "";
                     if (property.IsField)
                     {
-                        string fieldAccessorName = GetFieldAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+                        string fieldAccessorName = GetFieldAccessorName(declaringType, property.UnderlyingMemberName);
                         return $"{fieldAccessorName}({refPrefix}{objParam})";
                     }
                     else
                     {
-                        string propertyGetterAccessorName = GetPropertyGetterAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+                        string propertyGetterAccessorName = GetPropertyGetterAccessorName(declaringType, property.UnderlyingMemberName);
                         return $"{propertyGetterAccessorName}({refPrefix}{objParam})";
                     }
                 }
@@ -85,19 +85,19 @@ internal static partial class SourceFormatter
                 return $"{objParam}.{RoslynHelpers.EscapeKeywordIdentifier(property.UnderlyingMemberName)}";
             }
 
-            static string FormatSetterBody(string objParam, string valueParam, PropertyShapeModel property)
+            static string FormatSetterBody(string objParam, string valueParam, ObjectShapeModel declaringType, PropertyShapeModel property)
             {
                 if (!property.IsSetterAccessible)
                 {
                     string refPrefix = property.DeclaringType.IsValueType ? "ref " : "";
                     if (property.IsField)
                     {
-                        string fieldAccessorName = GetFieldAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+                        string fieldAccessorName = GetFieldAccessorName(declaringType, property.UnderlyingMemberName);
                         return $"{fieldAccessorName}({refPrefix}{objParam}) = {valueParam}";
                     }
                     else
                     {
-                        string propertySetterAccessorName = GetPropertySetterAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+                        string propertySetterAccessorName = GetPropertySetterAccessorName(declaringType, property.UnderlyingMemberName);
                         return $"{propertySetterAccessorName}({refPrefix}{objParam}, {valueParam})";
                     }
                 }
@@ -110,25 +110,25 @@ internal static partial class SourceFormatter
         writer.WriteLine("};");
     }
 
-    private static string GetFieldAccessorName(TypeId declaringType, string underlyingMemberName)
+    private static string GetFieldAccessorName(TypeShapeModel declaringType, string underlyingMemberName)
     {
-        return $"{declaringType.TypeIdentifier}_{underlyingMemberName}_Accessor";
+        return $"__FieldAccessor_{declaringType.SourceIdentifier}_{underlyingMemberName}";
     }
 
-    private static string GetPropertyGetterAccessorName(TypeId declaringType, string underlyingMemberName)
+    private static string GetPropertyGetterAccessorName(TypeShapeModel declaringType, string underlyingMemberName)
     {
-        return $"{declaringType.TypeIdentifier}_{underlyingMemberName}_GetAccessor";
+        return $"__GetAccessor_{declaringType.SourceIdentifier}_{underlyingMemberName}";
     }
 
-    private static string GetPropertySetterAccessorName(TypeId declaringType, string underlyingMemberName)
+    private static string GetPropertySetterAccessorName(TypeShapeModel declaringType, string underlyingMemberName)
     {
-        return $"{declaringType.TypeIdentifier}_{underlyingMemberName}_SetAccessor";
+        return $"__SetAccessor_{declaringType.SourceIdentifier}_{underlyingMemberName}";
     }
 
-    private static void FormatFieldAccessor(SourceWriter writer, PropertyShapeModel property)
+    private static void FormatFieldAccessor(SourceWriter writer, ObjectShapeModel declaringType, PropertyShapeModel property)
     {
         Debug.Assert(property.IsField);
-        string accessorName = GetFieldAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+        string accessorName = GetFieldAccessorName(declaringType, property.UnderlyingMemberName);
         string refPrefix = property.DeclaringType.IsValueType ? "ref " : "";
 
         writer.WriteLine($"""
@@ -137,10 +137,10 @@ internal static partial class SourceFormatter
             """);
     }
 
-    private static void FormatPropertyGetterAccessor(SourceWriter writer, PropertyShapeModel property)
+    private static void FormatPropertyGetterAccessor(SourceWriter writer, ObjectShapeModel declaringType, PropertyShapeModel property)
     {
         Debug.Assert(!property.IsField);
-        string accessorName = GetPropertyGetterAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+        string accessorName = GetPropertyGetterAccessorName(declaringType, property.UnderlyingMemberName);
         string refPrefix = property.DeclaringType.IsValueType ? "ref " : "";
         string propertyGetter = "get_" + property.UnderlyingMemberName;
 
@@ -148,10 +148,10 @@ internal static partial class SourceFormatter
         {
             writer.WriteLine($$"""
                 // Workaround for https://github.com/dotnet/runtime/issues/89439
-                private static global::System.Reflection.MethodInfo? s_{{accessorName}}_MethodInfo;
+                private static global::System.Reflection.MethodInfo? __s_{{accessorName}}_MethodInfo;
                 private static {{property.PropertyType.FullyQualifiedName}} {{accessorName}}({{refPrefix}}{{property.DeclaringType.FullyQualifiedName}} obj)
                 {
-                    global::System.Reflection.MethodInfo getter = s_{{accessorName}}_MethodInfo ??= typeof({{property.DeclaringType.FullyQualifiedName}}).GetMethod({{FormatStringLiteral(propertyGetter)}}, {{InstanceBindingFlagsConstMember}})!;
+                    global::System.Reflection.MethodInfo getter = __s_{{accessorName}}_MethodInfo ??= typeof({{property.DeclaringType.FullyQualifiedName}}).GetMethod({{FormatStringLiteral(propertyGetter)}}, {{InstanceBindingFlagsConstMember}})!;
                     return getter.Invoke(obj, null);
                 }
                 """);
@@ -165,10 +165,10 @@ internal static partial class SourceFormatter
             """);
     }
 
-    private static void FormatPropertySetterAccessor(SourceWriter writer, PropertyShapeModel property)
+    private static void FormatPropertySetterAccessor(SourceWriter writer, ObjectShapeModel declaringType, PropertyShapeModel property)
     {
         Debug.Assert(!property.IsField);
-        string accessorName = GetPropertySetterAccessorName(property.DeclaringType, property.UnderlyingMemberName);
+        string accessorName = GetPropertySetterAccessorName(declaringType, property.UnderlyingMemberName);
         string refPrefix = property.DeclaringType.IsValueType ? "ref " : "";
         string propertySetter = "set_" + property.UnderlyingMemberName;
 
@@ -176,13 +176,13 @@ internal static partial class SourceFormatter
         {
             writer.WriteLine($$"""
                 // Workaround for https://github.com/dotnet/runtime/issues/89439
-                private static global::System.Reflection.MethodInfo? s_{{accessorName}}_MethodInfo;
+                private static global::System.Reflection.MethodInfo? __s_{{accessorName}}_MethodInfo;
                 private static void {{accessorName}}({{refPrefix}}{{property.DeclaringType.FullyQualifiedName}} obj, {{property.PropertyType.FullyQualifiedName}} value)
                 {
                 """);
 
             writer.Indentation++;
-            writer.WriteLine($"global::System.Reflection.MethodInfo setter = s_{accessorName}_MethodInfo ??= typeof({property.DeclaringType.FullyQualifiedName}).GetMethod({FormatStringLiteral(propertySetter)}, {InstanceBindingFlagsConstMember})!;");
+            writer.WriteLine($"global::System.Reflection.MethodInfo setter = __s_{accessorName}_MethodInfo ??= typeof({property.DeclaringType.FullyQualifiedName}).GetMethod({FormatStringLiteral(propertySetter)}, {InstanceBindingFlagsConstMember})!;");
             if (property.DeclaringType.IsValueType)
             {
                 writer.WriteLine($$"""

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Type.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Type.cs
@@ -5,13 +5,13 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter
 {
-    private static SourceText FormatType(TypeShapeProviderModel provider, TypeShapeModel type)
+    private SourceText FormatType(TypeShapeProviderModel provider, TypeShapeModel type)
     {
         string generatedPropertyType = $"global::PolyType.Abstractions.ITypeShape<{type.Type.FullyQualifiedName}>";
-        string generatedFactoryMethodName = $"Create_{type.Type.TypeIdentifier}";
-        string generatedFieldName = "_" + type.Type.TypeIdentifier;
+        string generatedFactoryMethodName = $"__Create_{type.SourceIdentifier}";
+        string generatedFieldName = "__" + type.SourceIdentifier;
 
         var writer = new SourceWriter();
         StartFormatSourceFile(writer, provider.ProviderDeclaration);
@@ -22,7 +22,7 @@ internal static partial class SourceFormatter
 
         writer.WriteLine("/// <summary>Gets a generated shape for the specified type.</summary>");
         writer.WriteLine("#nullable disable annotations // Use nullable-oblivious property type", disableIndentation: true);
-        writer.WriteLine($"public {generatedPropertyType} {type.Type.TypeIdentifier} => {generatedFieldName} ??= {generatedFactoryMethodName}();");
+        writer.WriteLine($"public {generatedPropertyType} {type.SourceIdentifier} => {generatedFieldName} ??= {generatedFactoryMethodName}();");
         writer.WriteLine("#nullable enable annotations // Use nullable-oblivious property type", disableIndentation: true);
         writer.WriteLine($"private {generatedPropertyType}? {generatedFieldName};");
         writer.WriteLine();

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.cs
@@ -6,21 +6,29 @@ using PolyType.SourceGenerator.Model;
 
 namespace PolyType.SourceGenerator;
 
-internal static partial class SourceFormatter
+internal sealed partial class SourceFormatter(TypeShapeProviderModel provider)
 {
-    private const string InstanceBindingFlagsConstMember = "InstanceBindingFlags";
+    private const string InstanceBindingFlagsConstMember = "__BindingFlags_Instance_All";
+    private const string ProviderSingletonProperty = "Default";
+    private const string GetShapeMethodName = "GetShape";
+    public static string[] ReservedIdentifiers { get; } = [ProviderSingletonProperty,GetShapeMethodName];
 
-    public static void FormatProvider(SourceProductionContext context, TypeShapeProviderModel provider)
+    public static void GenerateSourceFiles(SourceProductionContext context, TypeShapeProviderModel provider)
+    {
+        SourceFormatter formatter = new(provider);
+        formatter.AddAllSourceFiles(context, provider);
+    }
+
+    private void AddAllSourceFiles(SourceProductionContext context, TypeShapeProviderModel provider)
     {
         context.CancellationToken.ThrowIfCancellationRequested();
-
         context.AddSource($"{provider.ProviderDeclaration.SourceFilenamePrefix}.g.cs", FormatMainFile(provider));
         context.AddSource($"{provider.ProviderDeclaration.SourceFilenamePrefix}.ITypeShapeProvider.g.cs", FormatProviderInterfaceImplementation(provider));
 
         foreach (TypeShapeModel type in provider.ProvidedTypes.Values)
         {
             context.CancellationToken.ThrowIfCancellationRequested();
-            context.AddSource($"{provider.ProviderDeclaration.SourceFilenamePrefix}.{type.Type.TypeIdentifier}.g.cs", FormatType(provider, type));
+            context.AddSource($"{provider.ProviderDeclaration.SourceFilenamePrefix}.{type.SourceIdentifier}.g.cs", FormatType(provider, type));
         }
 
         foreach (TypeDeclarationModel typeDeclaration in provider.AnnotatedTypes)
@@ -35,7 +43,7 @@ internal static partial class SourceFormatter
                 context.CancellationToken.ThrowIfCancellationRequested();
                 string sourceFile = typeToImplement == typeDeclaration.Id
                     ? $"{typeDeclaration.SourceFilenamePrefix}.IShapeable.g.cs"
-                    : $"{typeDeclaration.SourceFilenamePrefix}.IShapeable.{typeToImplement.TypeIdentifier}.g.cs";
+                    : $"{typeDeclaration.SourceFilenamePrefix}.IShapeable.{GetShapeModel(typeToImplement).SourceIdentifier}.g.cs";
                 context.AddSource(sourceFile, FormatIShapeableOfTStub(typeDeclaration, typeToImplement, provider));
             }
         }
@@ -57,7 +65,7 @@ internal static partial class SourceFormatter
                 global::System.Reflection.BindingFlags.Instance;
 
             /// <summary>Gets the default instance of the <see cref="{{provider.ProviderDeclaration.Name}}"/> class.</summary>
-            public static {{provider.ProviderDeclaration.Name}} Default { get; } = new();
+            public static {{provider.ProviderDeclaration.Name}} {{ProviderSingletonProperty}} { get; } = new();
 
             /// <summary>Initializes a new instance of the <see cref="{{provider.ProviderDeclaration.Name}}"/> class.</summary>
             public {{provider.ProviderDeclaration.Name}}() { }
@@ -111,6 +119,8 @@ internal static partial class SourceFormatter
             writer.WriteLine('}');
         }
     }
+
+    private TypeShapeModel GetShapeModel(TypeId typeId) => provider.ProvidedTypes[typeId];
 
     private static string FormatBool(bool value) => value ? "true" : "false";
     private static string FormatNull(string? stringExpr) => stringExpr is null ? "null" : stringExpr;

--- a/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.cs
+++ b/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.cs
@@ -327,4 +327,76 @@ public static class CompilationTests
         PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
         Assert.Empty(result.Diagnostics);
     }
+
+    [Fact]
+    public static void ConflictingTypeNames_NoErrors()
+    {
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+            using PolyType;
+
+            [GenerateShape]
+            public partial class MyPoco;
+
+            namespace Foo.Bar
+            {
+                partial class Container
+                {
+                    [GenerateShape]
+                    public partial class MyPoco;
+                }
+            }
+
+            namespace Foo.Baz
+            {
+                partial class Container
+                {
+                    [GenerateShape]
+                    public partial class MyPoco;
+                }
+            }
+            """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
+
+    [Fact]
+    public static void ConflictingTypeNamesInNestedGenerics_NoErrors()
+    {
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+            using PolyType;
+
+            class Container<T>
+            {
+                public partial class MyPoco;
+            }
+
+            [GenerateShape<Container<int>.MyPoco>]
+            [GenerateShape<Container<string>.MyPoco>]
+            partial class Witness;
+            """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
+
+    [Fact]
+    public static void TypesUsingReservedIdentifierNames_NoErrors()
+    {
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+            using PolyType;
+
+            [GenerateShape]
+            partial class Default;
+
+            [GenerateShape]
+            partial class GetShape;
+
+            [GenerateShape]
+            partial class @class;
+            """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
 }

